sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -406,11 +419,16 @@ object ScalaReflection extends ScalaReflection {
     def toCatalystArray(input: Expression, elementType: `Type`): Expression = {
       val externalDataType = dataTypeFor(elementType)
       val Schema(catalystType, nullable) = silentSchemaFor(elementType)
-      if (isNativeType(catalystType)) {
-        NewInstance(
+
+      if (isNativeType(catalystType) && !(elementType <:< localTypeOf[Option[_]])) {
```

Comment:
good catch!
The original one is wrong, how about `isNativeType(externalDataType)` which is simpler?